mgmtsystem_action/models/mgmtsystem_action.py

```diff
@@ -0,0 +1,223 @@
+
+    @api.model
+    def _get_stage_open(self):
+        return self.env.ref('mgmtsystem_action.stage_open')
+
+    @api.model
+    def _get_stage_close(self):
+        return self.env.ref('mgmtsystem_action.stage_close')
+
+    @api.model
+    def _get_stage_cancel(self):
+        return self.env.ref('mgmtsystem_action.stage_cancel')
+
+    @api.multi
+    def case_open(self):
+        """ Opens case """
+        for case in self:
+            case.write({
+                'active': True,
+                'stage_id': case._get_stage_open().id})
+        return True
+
+    @api.model
+    def create(self, vals):
+        """Creation of Action."""
+        Sequence = self.env['ir.sequence']
+        vals['reference'] = Sequence.next_by_code('mgmtsystem.action')
+        action = super(MgmtSystemAction, self).create(vals)
+        self.send_mail_for_action(action)
+        return action
+
+    @api.multi
+    def write(self, vals):
+        """Update user data."""
+        if vals.get('stage_id'):
+            stage_new = self._get_stage_new()
+            stage_open = self._get_stage_open()
+            stage_close = self._get_stage_close()
+            stage_cancel = self._get_stage_cancel()
+            if vals['stage_id'] == stage_new.id:
+                if self.opening_date:
+                    raise exceptions.ValidationError(
+                        _('We cannot bring back the action to draft stage')
+                    )
+                vals['cancel_date'] = None
+                self.message_post(
+                    body=' %s ' % (_('Action back to draft stage on ') +
+                                   fields.Datetime.now())
+                )
+            if vals['stage_id'] == stage_open.id:
+                vals['opening_date'] = fields.Datetime.now()
+                self.message_post(
+                    body=' %s ' % (_('Action opened on ') +
+                                   vals['opening_date'])
+                )
+                vals['date_closed'] = None
+                vals['cancel_date'] = None
+            if vals['stage_id'] == stage_close.id:
+                if not self.opening_date or self.cancel_date:
+                    raise exceptions.ValidationError(
+                        _('You should first open the action')
+                    )
+                vals['date_closed'] = fields.Datetime.now()
+                self.message_post(
+                    body=' %s ' % (_('Action closed on ') +
+                                   vals['date_closed'])
+                )
+            if vals['stage_id'] == stage_cancel.id:
+                vals['date_closed'] = None
+                vals['opening_date'] = None
+                vals['cancel_date'] = fields.Datetime.now()
+                self.message_post(
+                    body=' %s ' % (_('Action cancelled on ') +
+                                   fields.Datetime.now())
+                )
+        return super(MgmtSystemAction, self).write(vals)
```

Comment:
I strongly believe that there is a better way.
But that can be seen as an improvement, so can be left for a later PR.